### PR TITLE
Introduce the validation pipeline

### PR DESCRIPTION
**Pull request by CjS77**
The idea of ValidationPipeline is to provide a unified approach to
validation; and one that makes testing easier as we approach full
set of consensus rules.

A validation pipeline contains a set of validators that must all
pass before the pipeline passes.

The pipeline can maintain state (such as a BlockchainDB instance)
in order to help with validation.

There are also mock validators provided so that we can keep unit tests
of different parts of the consensus model separate, without
having to rewrite tests every time we add another consesnsus rule
in development.

This also lets us keep all validation in one place, making maintainability easier.

### CODE REVIEW

File: base_layer/core/src/validation/error.rs [hunk 1]
```diff
@@ -20,4 +20,40 @@
...
+    InvalidInput,
+    // Input maturity not reached
+    InputMaturity,
+    // Invalid coinbase maturity in block or more than one coinbase
+    InvalidCoinbase,
+}
+
+#[derive(Clone, Debug, PartialEq, Error)]
+pub enum BodyValidationError {
+    // The sum of the input and output commitments doesn't equal the sum of the kernel excesses
```
[the remark this answers is not in the document]
Comment (CjS77) [reply base_layer/core/src/validation/error.rs]: Gotcha, thx.

**Comment by CjS77** [comment]
This implementation is close, but not quite what we need. I've pushed some changes. Thoughts?

**Comment by CjS77** [comment]
Alternative is to split `Validator` into 3 traits.

Pro: Smaller binaries.
Con: You can't mix and match validators into a pipeline. TBH I can't think of a case where we'd actually want to run `validate()` in one and `validate_block()` in a single pipeline, so this might be a premature optimisation.

File: base_layer/core/src/validation/pipeline.rs [hunk 2]
```diff
@@ -0,0 +1,128 @@
...
+}
+
+impl Validation for ValidationPipeline {
+
+
+    fn validate(&mut self) -> Result<(), ValidationError> {
+        for v in &mut self.validators {
+            if let Err(e) = v.validate() {
+                return Err(e);
+            }
```
Comment (neonknight64) [inline base_layer/core/src/validation/pipeline.rs]: This if let can be replaced with `v.validate()?;`, a similar approach can be used for the `validate_block` and `validate_transaction` if let statements.

**Comment by sdbondi** [comment]
Just riffing here as I'm not sure exactly how this will be used.

```rust
trait Validator<T> {
   type Error: std::error::Error;

   fn validate(&mut self, item: &T) -> Result<(), Self::Error> {
      Ok(())
   }
}

struct SignatureValidator;
impl Validator<Transaction> for SignatureValidator {
   type Error = SignatureError;
  fn validate(&mut self, txn: &'a Transaction) -> Result<(), Self::Error> {
     validate_txn_sig(txn)
  }
}

impl Validator<Block> for SignatureValidator {
  fn validate(&mut self, block: & Block) -> Result<(), Self::Error> {
     validate_kernel_sigs(block)
  }
}
```

Would validators ever need to be async? 
```rust
// Async version for fun - though there will be lifetime issues
trait Validator<'a, T> {
   type Error: std::error::Error;
   type Future: Future<Output=Result<(), Self::Error>> + 'a;
 
   fn validate(&mut self, item: &'a T) -> Self::Future {
      future::ready(Ok(()));
   }
}

```

File: base_layer/core/src/validation/pipeline.rs (hunk 2 above)
Comment (CjS77) [reply base_layer/core/src/validation/pipeline.rs]: Yup, thanks.

File: base_layer/core/src/validation/pipeline.rs [hunk 3]
```diff
@@ -0,0 +1,128 @@
...
+        ValidationPipeline { validators: Vec::new() }
+    }
+}
+
+impl ValidationPipeline {
+    pub fn new() -> Self {
+        ValidationPipeline::default()
+    }
+
+    pub fn push(&mut self, validator: Pipelined) {
```
Comment (sdbondi) [inline base_layer/core/src/validation/pipeline.rs]: Boxing can also be done internally
```rust
    pub fn push<T>(&mut self, validator: T) {
      self.validators.push(Box::new(validator));
    }
```

**Comment by CjS77** [comment]
Hard to argue with this @sdbondi.

And good point on the `async`; I think in general we do want this to be async

The lifetime stuff is non-trivial, right? Doesn't the `'a` end up having to be static?

We could chicken out and have something like this maybe?

```
trait Validator<T> {
   type Error: std::error::Error;

   fn validate(&mut self, item: T) -> Result<T, Self::Error> {
      Ok(())
   }
}
```

**Comment by sdbondi** [comment]
I think if you use [`FuturesUnordered`](https://docs.rs/futures/0.3.1/futures/stream/futures_unordered/struct.FuturesUnordered.html) rather than spawning, the lifetimes may work out - but honestly, it's always a trial and error situation oftentimes resulting in 'chickening out' :) 

If the reference is kept, the implementation can always clone if required

**Comment by CjS77** [comment]
Incorporated @sdbondi's suggestions. 

Notes:
* The pipeline is synchronous. The idea is to pass the pipeline to a blocking thread in the scheduler.
* The `item`s are wrapped in an `Arc`. This makes them immutable without passing ownership over, which should preclude a bunch of bugs due to accidentally changing the state of the `item`s being validated.

File: base_layer/core/src/validation/pipeline.rs (hunk 3 above)
Comment (CjS77) [reply base_layer/core/src/validation/pipeline.rs]: Good tip. :+1:

File: base_layer/core/src/validation/error.rs [hunk 4]
```diff
@@ -20,4 +20,40 @@
...
 // USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.
 
-pub struct BlockValidationService;
+use derive_error::Error;
+use tari_transactions::transaction::TransactionError;
+
+#[derive(Clone, Debug, PartialEq, Error)]
+pub enum ValidationError {
+    BlockError(BlockValidationError),
+    BodyError(BodyValidationError),
```
Comment (SWvheerden) [inline base_layer/core/src/validation/error.rs]: cant we merge these two errors?

I see them as mostly the same.

**Pull request closed without merging**